In [2]:
# Initialize Otter
import otter
grader = otter.Notebook("lab02.ipynb")

<img style="display: block; margin-left: auto; margin-right: auto" src="./ccsf-logo.png" width="250rem;" alt="The CCSF black and white logo">

<div style="text-align: center;">
    <h1>Lab 02 - Data Types</h1>
    <em>View the related <a href="https://ccsf.instructure.com" target="_blank">Canvas</a> Assignment page for additional details.</em>
</div>

## References

* [datascience Documentation](https://datascience.readthedocs.io/)
* [Jupyter Notebook Documentation](https://jupyter-notebook.readthedocs.io/)
* [Markdown Cheat Sheet](https://www.markdownguide.org/cheat-sheet/)

First, run the following cell to set up the lab, and make sure you run the cell at the top of the notebook that initializes Otter.

In [ ]:
import numpy as np
from datascience import *

## Expressions Review

The two building blocks of Python code are *expressions* and *statements*.  An **expression** is a piece of code that

* is self-contained, meaning it would make sense to write it on a line by itself, and
* usually evaluates to a value.


Here are two expressions that both evaluate to 3:

    3
    5 - 2
    
One important type of expression is the **call expression**. A call expression begins with the name of a function and is followed by the argument(s) of that function in parentheses. The function returns some value, based on its arguments. Some important mathematical functions are listed below.

| Function | Description                                                   |
|----------|---------------------------------------------------------------|
| `abs`      | Returns the absolute value of its argument                    |
| `max`      | Returns the maximum of all its arguments                      |
| `min`      | Returns the minimum of all its arguments                      |
| `pow`      | Raises its first argument to the power of its second argument |
| `round`    | Rounds its argument to the nearest integer                     |

Here are two call expressions that both evaluate to 3:

    abs(2 - 5)
    max(round(2.8), min(pow(2, 10), -1 * pow(2, 10)))

The expression `2 - 5` and the two call expressions given above are examples of **compound expressions**, meaning that they are actually combinations of several smaller expressions.  `2 - 5` combines the expressions `2` and `5` by subtraction.  In this case, `2` and `5` are called **subexpressions** because they're expressions that are part of a larger expression.

A **statement** is a whole line of code.  Some statements are just expressions.  The expressions listed above are examples.

Other statements *make something happen* rather than *having a value*. For example, an **assignment statement** assigns a value to a name. 

A good way to think about this is that we're **evaluating the right-hand side** of the equals sign and **assigning it to the left-hand side**. Here are some assignment statements:
    
    height = 1.3
    the_number_five = abs(-5)
    absolute_height_difference = abs(height - 1.688)

An important idea in programming is that large, interesting things can be built by combining many simple, uninteresting things.  The key to understanding a complicated piece of code is breaking it down into its simple components.

For example, a lot is going on in the last statement above, but it's really just a combination of a few things.  This picture describes what's going on.

<img src="statement.png" alt="An assignment statement ">

### Task 01 📍

In the next cell, assign the name `max_value` to the larger value among the following two expressions:

1. the **absolute value** of $2^{5}-2^{11}-2^3 + 1$, and 
2. $5 \times 13 \times 31 + 7$.

Don't do this in one line. Instead, break this task up into steps. For example:

1. Assign `value_1` to the value associated with the first expression.
2. Assign `value_2` to the value associated with the second expression.
3. Use `max` to find the bigger of the two values.

In [4]:
value_1 = abs(2**5 - 2**11 - 2**3 + 1)
value_2 = 5 * 13 * 31 + 7
max_value = max(value_1, value_2)
max_value

2023

In [5]:
grader.check("task_01")

task_01 results: All test cases passed!
task_01 - 1 message: ✅ Great work running an assignment statement with the name max_value.
task_01 - 2 message: ✅ Great work! Hopefully, you were able to utilize the max and abs functions.

It is tempting to try and write the solution to that task in one line of code, but it is not a good idea to jam these steps into a single line because it can make the code harder to read and harder to debug. Good programming practice involves splitting up your code into smaller steps and using appropriate names. You'll have plenty of practice in the rest of this course!

## Text

Programming doesn't just concern numbers. Text is one of the most common data types used in programs. 

Text is represented by a **string value** in Python. The word "string" is a programming term for a sequence of characters. A string might contain a single character, a word, a sentence, or a whole book.

To distinguish text data from actual code, we demarcate strings by putting quotation marks around them. Single quotes (`'`) and double quotes (`"`) are both valid, but the types of opening and closing quotation marks must match. The contents can be any sequence of characters, including numbers and symbols. 

We've seen strings before in `print` statements.  Below, two different strings are passed as arguments to the `print` function.

In [6]:
print("I <3", 'Data Science')

I <3 Data Science


Just as names can be given to numbers, names can be given to string values.  The names and strings aren't required to be similar in any way. Any name can be assigned to any string.

In [7]:
one = '1'
plus = '+'
print(one, plus, one)

1 + 1


### Task 02 📍

Yuri Gagarin was the first person to travel through outer space.  When he emerged from his capsule upon landing on Earth, he [reportedly](https://en.wikiquote.org/wiki/Yuri_Gagarin) had the following conversation with a woman and girl who saw the landing:

    The woman asked: "Can it be that you have come from outer space?"
    Gagarin replied: "As a matter of fact, I have!"

The cell below contains unfinished code.  Fill in the `...`s so that it prints out this conversation *exactly* as it appears above. To complete this task successfully, the capitalization, spaces, punctuation, etc. all need to be correct.

In [8]:
woman_asking = 'The woman asked:'
woman_quote = '"Can it be that you have come from outer space?"'
gagarin_reply = 'Gagarin replied:'
gagarin_quote = '"As a matter of fact, I have!"'

print(woman_asking, woman_quote)
print(gagarin_reply, gagarin_quote)

The woman asked: "Can it be that you have come from outer space?"
Gagarin replied: "As a matter of fact, I have!"


In [9]:
grader.check("task_02")

task_02 results: All test cases passed!
task_02 - 1 message: ✅ It seems like you have the correct string.
task_02 - 2 message: ✅ It seems like you have the correct string.

### String Methods

Strings can be transformed using **methods**. Recall that methods and functions are not technically the same thing, but we'll be using them interchangeably for the purposes of this course.

Here's a sketch of how to call methods on a string:

    <expression that evaluates to a string>.<method name>(<argument>, <argument>, ...)
    
One example of a string method is `replace`, which replaces all instances of some part of the original string (or a *substring*) with a new string. 

    <original string>.replace(<old substring>, <new substring>)
    
`replace` returns (evaluates to) a new string, leaving the original string unchanged.
    
Try to predict the output of this example, then run the cell!

In [1]:
# Replace one letter
hello = 'Hello'
print(hello.replace('o', 'a'), hello)

Hella Hello


You can call functions on the results of other functions.  For example, `max(abs(-5), abs(3))` evaluates to 5.  Similarly, you can call methods on the results of other method or function calls.

You may have already noticed one difference between functions and methods - a function like `max` does not require a `.` before it's called, but a string method like `replace` does.

In [2]:
# Calling replace on the output of another call to replace
'train'.replace('t', 'ing').replace('in', 'de')

'degrade'

Here's a picture of how Python evaluates a "chained" method call like that:

<img src="./chaining_method_calls.png" alt="showing chaining methods" width=80%/>

#### Task 03 📍

Use `replace` to transform the string `'hitchhiker'` into `'matchmaker'`. Assign your result to `new_word`.


In [10]:
new_word = 'hitchhiker'.replace('hitch', 'match').replace('hiker', 'maker')
new_word

'matchmaker'

In [11]:
grader.check("task_03")

task_03 results: All test cases passed!
task_03 - 1 message: ✅ It seems like you have the correct string.

There are many more string methods in Python, but most programmers don't memorize their names or how to use them.  In the "real world," people usually just search the internet for documentation and examples. A complete [list of string methods](https://docs.python.org/3/library/stdtypes.html#string-methods) appears in the Python language documentation.

### Converting To and From Strings

Strings and numbers are different *types* of values, even when a string contains the digits of a number. For example, evaluating the following cell causes an error because an integer cannot be added to a string.

<img src="./int_add_str_error.png">

However, there are built-in functions to convert numbers to strings and strings to numbers. Some of these built-in functions have restrictions on the type of argument they take:

|Function |Description|
|-|-|
|`int`|Converts a string of digits or a float to an integer ("int") value|
|`float`|Converts a string of digits (perhaps with a decimal point) or an int to a decimal ("float") value|
|`str`|Converts any value to a string|

Try to predict what data type and value `example` evaluates to, then run the cell.

In [12]:
example = 8 + int("10") + float("8")

print(example)
print("This example returned a " + str(type(example)) + "!")

26.0
This example returned a <class 'float'>!


Suppose you're writing a program that looks for dates in a text, and you want your program to find the amount of time that elapsed between two years it has identified.  It doesn't make sense to subtract two texts, but you can first convert the text containing the years into numbers.

#### Task 04 📍

Finish the code below to compute the number of years that elapsed between `one_year` and `another_year`.  Don't just write the numbers `1618` and `1648` (or `30`); use a conversion function to turn the given text data into numbers.


In [13]:
# Some text data:
one_year = "1618"
another_year = "1648"

# Complete the next line.  Note that we can't just write:
#   another_year - one_year
# If you don't see why, try seeing what happens when you
# write that here.
difference = int(another_year) - int(one_year)
difference

30

In [14]:
grader.check("task_04")

task_04 results: All test cases passed!
task_04 - 1 message: ✅ It seems like you have the correct value.

### Passing strings to functions

String values, like numbers, can be arguments to functions and can be returned by functions. 

The function `len` (derived from the word "length") takes a single string as its argument and returns the number of characters (including spaces) in the string.

Note that it doesn't count *words*. `len("one small step for man")` evaluates to 22, not 5.

#### Task 05 📍

Use `len` to find the number of characters in the long string in the next cell.  Characters include things like spaces and punctuation. Assign `sentence_length` to that number.

(The string is the first sentence of the English translation of the French [Declaration of the Rights of Man](http://avalon.law.yale.edu/18th_century/rightsof.asp).)  


In [15]:
a_very_long_sentence = "The representatives of the French people, organized as a National Assembly, believing that the ignorance, neglect, or contempt of the rights of man are the sole cause of public calamities and of the corruption of governments, have determined to set forth in a solemn declaration the natural, unalienable, and sacred rights of man, in order that this declaration, being constantly before all the members of the Social body, shall remind them continually of their rights and duties; in order that the acts of the legislative power, as well as those of the executive power, may be compared at any moment with the objects and purposes of all political institutions and may thus be more respected, and, lastly, in order that the grievances of the citizens, based hereafter upon simple and incontestable principles, shall tend to the maintenance of the constitution and redound to the happiness of all."
sentence_length = len(a_very_long_sentence)
sentence_length

896

In [16]:
grader.check("task_05")

task_05 results: All test cases passed!
task_05 - 1 message: ✅ That seems correct!

## Importing Code

Most programming involves work that is very similar to work that has been done before.  Since writing code is time-consuming, it's good to rely on others' published code when you can.  Rather than copy-pasting, Python allows us to **import modules**. A module is a file with Python code that has defined variables and functions. By importing a module, we are able to use its code in our own notebook.

Python includes many useful modules that are just an `import` away.  We'll look at the `math` module as a first example. The `math` module is extremely useful in computing mathematical expressions in Python. 

Suppose we want to very accurately compute the area of a circle with a radius of 5 meters.  For that, we need the constant $\pi$, which is roughly 3.14.  Conveniently, the `math` module has `pi` defined for us:

In [17]:
import math
radius = 5
area_of_circle = radius ** 2 * math.pi
area_of_circle

78.53981633974483

In the code above, the line `import math` imports the math module. This statement creates a module and then assigns the name `math` to that module. We are now able to access any variables or functions defined within `math` by typing the name of the module followed by a dot, then followed by the name of the variable or function we want.

    <module name>.<name>

### Task 06 📍

The module `math` also provides the name `e` for the base of the natural logarithm, which is roughly 2.71. Compute $e^{\pi}-\pi$, giving it the name `near_twenty`.

**Hint**: `math.pi` and `math.e` provide representations of the mathematical constants $\pi$ and $e$.

In [18]:
near_twenty = math.exp(math.pi) - math.pi
near_twenty

19.999099979189474

In [19]:
grader.check("task_06")

task_06 results: All test cases passed!
task_06 - 1 message: ✅ Great work running the cell above that imports the math module.
task_06 - 2 message: ✅ Great work. It seems you have correct value.

<img src="./e_to_the_pi_minus_pi.png" alt="A comic about e to the pi minus pi.">

[An explanation of the comic.](https://www.explainxkcd.com/wiki/index.php/217:_e_to_the_pi_Minus_pi)

### Accessing functions

In the question above, you accessed variables within the `math` module. 

**Modules** also define **functions**.  For example, `math` provides the name `sqrt` for the square root function.  Having imported `math` already, we can write `math.sqrt(3)` to compute the square root of 3.

#### Task 07 📍

What is the square root of pi?  Compute that value using `sqrt` and `pi` from the `math` module.  Give the result the name `square_root_of_pi`.

In [20]:
square_root_of_pi = math.sqrt(math.pi)
square_root_of_pi

1.7724538509055159

In [21]:
grader.check("task_07")

task_07 results: All test cases passed!
task_07 - 1 message: ✅ Great work. It seems you have correct value.

Different functions can take in different numbers of arguments. Often, the [documentation](https://docs.python.org/3/library/math.html) of the module will provide information on how many arguments are required for each function.  

For example, the logarithm function requires two arguments, as shown below. Calculating logarithms (the logarithm of 8 in base 2). Remember that logarithms are the inverse of powers, so the result is 3 because 2 to the power of 3 is 8.

*Hint: If you press `shift+tab` while next to the function call, the documentation for that function will appear.*

In [31]:
math.log(8, 2)

3.0

There are various ways to import and access code from outside sources. The method we used above — `import <module_name>` — imports the entire module and requires that we use `<module_name>.<name>` to access its code. 

We can also import a specific constant or function instead of the entire module. Notice that you don't have to use the module name beforehand to reference that particular value. However, you do have to be careful about reassigning the names of the constants or functions to other values!

Importing just `sqrt` and `pi` from math means that we don't have to use `math.` in front of `sqrt` or `pi`.

In [32]:
from math import sqrt, pi
print(sqrt(pi))

1.7724538509055159


We do have to use it in front of other functions from math that we have not imported directly.

In [33]:
math.log(pi)

1.1447298858494002

We can also import **every** function and value from the entire module. This isn't a good practice in general, but we will do it in this class with the `datascience` module. Notice that, with the following command, you can just use `log` instead of `math.log`.

In [34]:
from math import *
log(pi)

1.1447298858494002

For our class, it is not important to focus on which type of import to use. In almost every case, there will be a cell containing all the import commands you need to run to complete the assignments.

## Arrays

Computers are most useful when you can use a small amount of code to *do the same action* to *many different things*.

For example, in the time it takes you to calculate the 18% tip on a restaurant bill, a laptop can calculate 18% tips for every restaurant bill paid by every human on Earth that day.  (That's if you're pretty fast at doing arithmetic in your head!)

**Arrays** are how we put many values in one place so that we can operate on them as a group. For example, if `billions_of_numbers` is an array of numbers, the expression

    .18 * billions_of_numbers

gives a new array of numbers that contains the result of multiplying each number in `billions_of_numbers` by .18.  Arrays are not limited to numbers; we can also put all the words in a book into an array of strings.

Concretely, an array is a **collection of values of the same type**. 

### Making arrays

First, let's learn how to manually input values into an array. This typically isn't how programs work. Normally, we create arrays by loading them from an external source, like a data file.

To create an array by hand, call the function `make_array`.  Each argument you pass to `make_array` will be in the array it returns.  Run this cell to see an example:

In [37]:
make_array(0.125, 4.75, -1.3)

array([ 0.125,  4.75 , -1.3  ])

Each value in an array (in the above case, the numbers 0.125, 4.75, and -1.3) is called an *element* of that array.

Arrays themselves are also values, just like numbers and strings.  That means you can assign them to names or use them as arguments to functions. For example, `len(<some_array>)` returns the number of elements in `some_array`.

#### Task 08 📍

Make an array containing the numbers 0, 1, -1, $\pi$, and $e$, in that order.  Name it `interesting_numbers`.  

*Hint:* How did you get the values $\pi$ and $e$ in Task 06?  You can refer to them in exactly the same way here.


In [22]:
interesting_numbers = make_array(0, 1, -1, math.pi, math.e)
interesting_numbers

array([ 0.        ,  1.        , -1.        ,  3.14159265,  2.71828183])

In [23]:
grader.check("task_08")

task_08 results: All test cases passed!
task_08 - 1 message: ✅ It seems like you made an array.
task_08 - 2 message: ✅ You have the correct number of values.
task_08 - 3 message: ✅ It seems like you have the correct values.

#### Task 09 📍

Make an array containing the five strings `"Hello"`, `","`, `" "`, `"world"`, and `"!"`.  (The third one is a single space inside quotes.)  Name it `hello_world_components`.

*Note:* If you evaluate `hello_world_components`, you'll notice some extra information in addition to its contents: `dtype='<U5'`.  That's just NumPy's extremely cryptic way of saying that the data types in the array are strings.


In [28]:
hello_world_components = make_array("Hello", ",", " ", "world", "!")
hello_world_components

array(['Hello', ',', ' ', 'world', '!'],
      dtype='<U5')

In [29]:
grader.check("task_09")

task_09 results: All test cases passed!
task_09 - 1 message: ✅ It seems like you made an array.
task_09 - 2 message: ✅ You have the correct number of values.
task_09 - 3 message: ✅ It looks like your array contains the blank space!
task_09 - 4 message: ✅ It seems like you have the correct values.

####  `np.arange`

Arrays are provided by a package called [NumPy](http://www.numpy.org/) (pronounced "NUM-pie"). The package is called `numpy`, but it's standard to rename it `np` for brevity.  You can do that with:

    import numpy as np

Very often in data science, we want to work with many numbers that are evenly spaced within some range.  NumPy provides a special function for this called `arange`.  The line of code `np.arange(start, stop, step)` evaluates to an array with all the numbers starting at `start` and counting up by `step`, stopping **before** `stop` is reached.

Note that this is not exactly the same as the built in data type `range`!

Run the following cells to see some examples!

In [45]:
# This array starts at 1 and counts up by 2
# and then stops before 6
np.arange(1, 6, 2)

array([1, 3, 5])

In [46]:
# This array doesn't contain 9
# because np.arange stops *before* the stop value is reached
np.arange(4, 9, 1)

array([4, 5, 6, 7, 8])

#### Task 10 📍

Use `np.arange` to create an array with the multiples of 99 from 0 up to (**and including**) 9999.  (So its elements are 0, 99, 198, 297, etc.)


In [34]:
multiples_of_99 = np.arange(0, 10000, 99)
multiples_of_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

In [35]:
grader.check("task_10")

task_10 results: All test cases passed!
task_10 - 1 message: ✅ It seems like you made an array.
task_10 - 2 message: ✅ You have the correct number of values.
task_10 - 3 message: ✅ It seems like you have the correct values.

### Indexing

Let's work with a more interesting dataset.  The next cell creates an array called `population_amounts` that includes estimated world populations in every year from **1950** to roughly the present.  (The estimates come from the US Census Bureau website.)

Rather than type in the data manually, we've loaded them from a file on your computer called `world_population.csv`.  You'll learn about this code later.

In [37]:
population_amounts = Table.read_table("./world_population.csv").column("Population")
population_amounts

array([2557628654, 2594939877, 2636772306, 2682053389, 2730228104,
       2782098943, 2835299673, 2891349717, 2948137248, 3000716593,
       3043001508, 3083966929, 3140093217, 3209827882, 3281201306,
       3350425793, 3420677923, 3490333715, 3562313822, 3637159050,
       3712697742, 3790326948, 3866568653, 3942096442, 4016608813,
       4089083233, 4160185010, 4232084578, 4304105753, 4379013942,
       4451362735, 4534410125, 4614566561, 4695736743, 4774569391,
       4856462699, 4940571232, 5027200492, 5114557167, 5201440110,
       5288955934, 5371585922, 5456136278, 5538268316, 5618682132,
       5699202985, 5779440593, 5857972543, 5935213248, 6012074922,
       6088571383, 6165219247, 6242016348, 6318590956, 6395699509,
       6473044732, 6551263534, 6629913759, 6709049780, 6788214394,
       6866332358, 6944055583, 7022349283, 7101027895, 7178722893,
       7256490011])

Run the following cell to confirm that `population_amounts` is a NumPy array.

In [53]:
type(population_amounts)

numpy.ndarray

Here's how we get the first element of `population_amounts`, which is the world population in the first year in the dataset, 1950.

In [54]:
population_amounts.item(0)

2557628654

The value of that expression is the number 2557628654 (around 2.5 billion), because that's the first thing in the array `population_amounts`.

Notice that we wrote `.item(0)`, not `.item(1)`, to get the first element.  This is a weird convention in computer science.  0 is called the *index* of the first item.  It's the number of elements that appear *before* that item.  So 3 is the index of the 4th item.

**Note:** It is possible to use `population_amounts[0]` to get the first element, but the course material does not guide you this way, so be careful and always check the data type of your results. The `.item` method from the `datascience` module is performing some additional code that the square bracket notation is not performing.

In [55]:
type(population_amounts.item(0))

int

In [56]:
type(population_amounts[0])

numpy.int64

Here are some more examples.  In the examples, we've given names to the things we get out of `population_amounts`.  Read and run each cell.

In [57]:
# The 13th element in the array is the population
# in 1962 (which is 1950 + 12).
population_1962 = population_amounts.item(12)
population_1962

3140093217

In [58]:
# The 66th element is the population in 2015.
population_2015 = population_amounts.item(65)
population_2015

7256490011

You will see an `IndexError` if you try to use an index value that is outside the range of possible values for the array:

<img src="./index_error.png">

Since `make_array` returns an array, we can call `.item(3)` on its output to get its 4th element, just like we "chained" together calls to the method `replace` earlier.

In [59]:
make_array(-1, -3, 4, -2).item(3)

-2

#### Task 11 📍

Set `population_1973` to the world population in 1973, by getting the appropriate element from `population_amounts` using `item`.


In [38]:
population_1973 = population_amounts.item(23)
population_1973

3942096442

In [39]:
grader.check("task_11")

task_11 results: All test cases passed!
task_11 - 1 message: ✅ It seems like you used the .item method.
task_11 - 2 message: ✅ You have the correct value.

Arrays are primarily useful for doing the same operation many times, so we don't often have to use `.item` and work with single elements.

### Logarithms

Here is one question we might ask about world population:

> How big was the population in *orders of magnitude* in each year?

Orders of magnitude quantify how big a number is by representing it as the power of another number (for example, representing $104$ as $10^{2.017033}$). One way to do this is by using the logarithm function. The logarithm (base 10) of a number increases by 1 every time we multiply the number by 10. It's like a measure of how many decimal digits the number has, or how big it is in orders of magnitude.

We can use the `log10` function from the `math` module to see that $1004$ is an order of magnitude larger than $104$. 

In [70]:
math.log10(104)

2.0170333392987803

In [71]:
math.log10(1004)

3.0017337128090005

To answer the original question, we could try to answer our question like this, using the `log10` function from the `math` module and the `item` method you just saw:

In [ ]:
population_1950_magnitude = math.log10(population_amounts.item(0))
population_1951_magnitude = math.log10(population_amounts.item(1))
population_1952_magnitude = math.log10(population_amounts.item(2))
population_1953_magnitude = math.log10(population_amounts.item(3))
# and so on

But this is tedious and doesn't really take advantage of the fact that we are using a computer.

Instead, NumPy provides its own version of `log10` that takes the logarithm of each element of an array.  It takes a single array of numbers as its argument.  It returns an array of the same length, where the first element of the result is the logarithm of the first element of the argument, and so on.

For example:

In [72]:
np.log10(104)

2.0170333392987803

In [73]:
an_array = make_array(104, 1004)
np.log10(an_array)

array([ 2.01703334,  3.00173371])

#### Task 12 📍

Use `np.log10` to compute the logarithms of the world population in every year.  Give the result (an array of 66 numbers) the name `population_magnitudes`.  Your code should be very short.


In [40]:
population_magnitudes = np.log10(population_amounts)
population_magnitudes

array([ 9.40783749,  9.4141273 ,  9.42107263,  9.42846742,  9.43619893,
        9.44437257,  9.45259897,  9.46110062,  9.4695477 ,  9.47722498,
        9.48330217,  9.48910971,  9.49694254,  9.50648175,  9.51603288,
        9.5251    ,  9.53411218,  9.54286695,  9.55173218,  9.56076229,
        9.56968959,  9.57867667,  9.58732573,  9.59572724,  9.60385954,
        9.61162595,  9.61911264,  9.62655434,  9.63388293,  9.64137633,
        9.64849299,  9.6565208 ,  9.66413091,  9.67170374,  9.67893421,
        9.68632006,  9.69377717,  9.70132621,  9.70880804,  9.7161236 ,
        9.72336995,  9.73010253,  9.73688521,  9.74337399,  9.74963446,
        9.75581413,  9.7618858 ,  9.76774733,  9.77343633,  9.77902438,
        9.7845154 ,  9.78994853,  9.7953249 ,  9.80062024,  9.80588805,
        9.81110861,  9.81632507,  9.82150788,  9.82666101,  9.83175555,
        9.83672482,  9.84161319,  9.84648243,  9.85132122,  9.85604719,
        9.8607266 ])

In [41]:
grader.check("task_12")

task_12 results: All test cases passed!
task_12 - 1 message: ✅ It seems like you made an array.
task_12 - 2 message: ✅ It seems like you have the correct values.

What you just did is called *elementwise* application of `np.log10`, since `np.log10` operates separately on each element of the array that it's called on. Here's a picture of what's going on:

<img src="./array_logarithm.jpg">


The textbook's [section](https://www.inferentialthinking.com/chapters/05/1/Arrays)  on arrays has a useful list of NumPy functions that are designed to work elementwise, like `np.log10`.

### Arithmetic

Arithmetic also works elementwise on arrays, meaning that if you perform an arithmetic operation (like subtraction, division, etc) on an array, Python will do the operation to every element of the array individually and return an array of all of the results. For example, you can divide all the population numbers by 1 billion to get numbers in billions:

In [ ]:
population_in_billions = population_amounts / 1000000000
population_in_billions

You can do the same with addition, subtraction, multiplication, and exponentiation (`**`). For example, you can calculate a tip on several restaurant bills at once (in this case just 3):

In [78]:
restaurant_bills = make_array(20.12, 39.90, 31.01)
print("Restaurant bills:\t", restaurant_bills)

# Array multiplication
tips = .2 * restaurant_bills
print("Tips:\t\t\t", tips)

Restaurant bills:	 [ 20.12  39.9   31.01]
Tips:			 [ 4.024  7.98   6.202]


<img src="./array_multiplication.jpg">

Suppose the total charge at a restaurant is the original bill plus the tip. If the tip is 20%, that means we can multiply the original bill by 1.2 to get the total charge.  Run the following code to calculate the total charge for each bill in `restaurant_bills`, and assign the resulting array to `total_charges`.

In [84]:
total_charges = 1.2 * restaurant_bills
total_charges

array([ 24.144,  47.88 ,  37.212])

#### Task 13 📍

The array `more_restaurant_bills` contains 100,000 bills!  Compute the total charge for every bill.

*Note: You'll learn about the command `Table.read_table` soon. For now, just run the cell.*


In [1]:
more_restaurant_bills = Table.read_table("./more_restaurant_bills.csv").column("Bill")
more_total_charges = np.array(more_restaurant_bills) * 1.2
more_total_charges

NameError: name 'Table' is not defined

In [43]:
grader.check("task_13")

task_13 results: All test cases passed!
task_13 - 1 message: ✅ It seems like you made an array.
task_13 - 2 message: ✅ It seems like you have the correct values.

The function `np.sum` takes a single array of numbers as its argument.  It returns the sum of all the numbers in that array (so it returns a single number, not an array).

#### Task 14 📍

What was the sum of all the bills in `more_restaurant_bills`, *including tips*?


In [44]:
sum_of_bills = sum(np.array(more_restaurant_bills) * 1.2)
sum_of_bills

1795730.0640000193

In [45]:
grader.check("task_14")

task_14 results: All test cases passed!
task_14 - 1 message: ✅ It seems like you have the correct values.

#### Task 15 📍

The powers of 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) arise frequently in computer science.  (For example, you may have noticed that storage on smartphones or USBs come in powers of 2, like 16 GB, 32 GB, or 64 GB.)  

Use the exponentiation operator `**` with `np.arange` to compute the first 30 powers of 2, starting from $2^0$ up to and including $2^{29}$.

* *Hint 1:* `2 ** make_array(1, 2)` would produce the array `array([2, 4])`.
* *Hint 2:* `np.arange(1, 2**30, 1)` creates an array with $2^{30}$ elements and **will crash your kernel**.



In [50]:
powers_of_2 = 2 ** np.arange(30)
powers_of_2

array([        1,         2,         4,         8,        16,        32,
              64,       128,       256,       512,      1024,      2048,
            4096,      8192,     16384,     32768,     65536,    131072,
          262144,    524288,   1048576,   2097152,   4194304,   8388608,
        16777216,  33554432,  67108864, 134217728, 268435456, 536870912])

In [51]:
grader.check("task_15")

task_15 results: All test cases passed!
task_15 - 1 message: ✅ It seems like you have the correct values.

## Submit your Lab to Canvas

Once you have finished working on the lab questions, prepare to submit your work in Canvas by completing the following steps.

1. In the related Canvas Assignment page, check the requirements for a Complete score for this lab assignment.
2. Double-check that you have run the code cell near the end of the notebook that contains the command `grader.check_all()`. This command will run all of the run tests on all your responses to the auto-graded tasks marked with 📍.
3. Double-check your responses to the manually graded tasks marked with 📍🔎.
4. Select the menu items `File`, `Save and Export Notebook As...`, and `Html_embed` in the notebook's Toolbar to download an HTML version of this notebook file.
5. In the related Canvas Assignment page, click Start Assignment or New Attempt to upload the downloaded HTML file.

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [52]:
grader.check_all()

task_01 results: All test cases passed!
task_01 - 1 message: ✅ Great work running an assignment statement with the name max_value.
task_01 - 2 message: ✅ Great work! Hopefully, you were able to utilize the max and abs functions.

task_02 results: All test cases passed!
task_02 - 1 message: ✅ It seems like you have the correct string.
task_02 - 2 message: ✅ It seems like you have the correct string.

task_03 results: All test cases passed!
task_03 - 1 message: ✅ It seems like you have the correct string.

task_04 results: All test cases passed!
task_04 - 1 message: ✅ It seems like you have the correct value.

task_05 results: All test cases passed!
task_05 - 1 message: ✅ That seems correct!

task_06 results: All test cases passed!
task_06 - 1 message: ✅ Great work running the cell above that imports the math module.
task_06 - 2 message: ✅ Great work. It seems you have correct value.

task_07 results: All test cases passed!
task_07 - 1 message: ✅ Great work. It seems you have correct val